nmpolicy/internal/resolver/filter.go

```diff
@@ -48,19 +41,78 @@ func filter(inputState map[string]interface{}, path ast.VariadicOperator, expect
 	return filteredMap, nil
 }
 
-func mapContainsValue(expectedValue interface{}) mapEntryVisitFn {
-	return func(mapToFilter map[string]interface{}, filterKey string) (interface{}, error) {
-		obtainedValue, ok := mapToFilter[filterKey]
-		if !ok {
-			return nil, nil
+type eqFilterVisitor struct {
+	mergeVisitResult bool
+	expectedValue    interface{}
+}
+
+func (e *eqFilterVisitor) visitLastMap(p path, mapToFilter map[string]interface{}) (interface{}, error) {
+	obtainedValue, ok := mapToFilter[*p.currentStep.Identity]
+	if !ok {
+		return nil, nil
+	}
+	if reflect.TypeOf(obtainedValue) != reflect.TypeOf(e.expectedValue) {
+		return nil, pathError(p.currentStep, `type missmatch: the value in the path doesn't match the value to filter. `+
+			`"%T" != "%T" -> %+v != %+v`, obtainedValue, e.expectedValue, obtainedValue, e.expectedValue)
+	}
+	if obtainedValue == e.expectedValue {
+		return mapToFilter, nil
+	}
+	return nil, nil
+}
+
+func (e *eqFilterVisitor) visitLastSlice(p path, _ []interface{}) (interface{}, error) {
+	return nil, nil
+}
+
+func (e *eqFilterVisitor) visitMap(p path, mapToVisit map[string]interface{}) (interface{}, error) {
+	interfaceToVisit, ok := mapToVisit[*p.currentStep.Identity]
+	if !ok {
+		return nil, nil
+	}
+	visitResult, err := visitState(p.nextStep(), interfaceToVisit, e)
+	if err != nil {
+		return nil, err
+	}
+	if visitResult == nil {
+		return nil, nil
+	}
+	filteredMap := map[string]interface{}{}
+	if e.mergeVisitResult {
+		for k, v := range mapToVisit {
+			filteredMap[k] = v
 		}
-		if reflect.TypeOf(obtainedValue) != reflect.TypeOf(expectedValue) {
-			return nil, fmt.Errorf(`type missmatch: the value in the path doesn't match the value to filter. `+
-				`"%T" != "%T" -> %+v != %+v`, obtainedValue, expectedValue, obtainedValue, expectedValue)
+	}
+	filteredMap[*p.currentStep.Identity] = visitResult
+	return filteredMap, nil
+}
+
+func (e *eqFilterVisitor) visitSlice(p path, sliceToVisit []interface{}) (interface{}, error) {
+	if p.currentStep.Number != nil {
```

Comment:
I think both `visitSlice` and `visitMap` should return an error if identity is nil.

Reply:
You mean with visitSlice it should be a Numeber and with visitMap it should be a identity right ?

Reply:
Make sense, let's throw an error.